### PR TITLE
feat: spu-to-spu mirroring connection

### PR DESCRIPTION
**Pull request by fraidev**
Needs: 
- [x] https://github.com/infinyon/fluvio/pull/3979

### CODE REVIEW

File: crates/fluvio-controlplane/src/spu_api/update_mirror.rs [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+pub type MirrorMsg = Message<Mirror>;
+pub type MirrorMsgs = Messages<Mirror>;
+
+impl<C> From<MetadataStoreObject<MirrorSpec, C>> for Mirror
+where
+    C: MetadataItem,
+{
+    fn from(mso: MetadataStoreObject<MirrorSpec, C>) -> Self {
+        let name = mso.key_owned();
```
Comment (galibey) [inline crates/fluvio-controlplane/src/spu_api/update_mirror.rs:37]: No need to call `key_owned` which does clone inside, you can just move `key` field out, it is public

File: crates/fluvio-spu/src/services/public/mod.rs [hunk 2]
```diff
@@ -76,100 +78,119 @@ impl FluvioService for PublicService {
...
-                    break;
-                }
-                None => {
-                    debug!(sink_id = shared_sink.id(), "No content, end of connection",);
-                    break;
                 }
             }
         }
 
+        if let Some(request) = mirror_request {
```
Comment (galibey) [inline crates/fluvio-spu/src/services/public/mod.rs:190]: Why can't we respond in place? Why saving the request to var and process outside the loop?

File: crates/fluvio-spu/src/mirroring/remote/controller.rs [hunk 3]
```diff
@@ -0,0 +1,523 @@
...
+    }
+
+    /// create socket to home, this will always succeed
+    #[instrument]
+    async fn create_socket_to_home(
+        &self,
+        backoff: &mut ExponentialBackoff,
+        _home: &Home,
+    ) -> (FluvioSocket, bool) {
+        //TODO: remove the comments bellow when we have tls support
```
Comment (galibey) [inline crates/fluvio-spu/src/mirroring/remote/controller.rs:437]: We should not use comments as code storage.

Comment (fraidev) [reply crates/fluvio-spu/src/mirroring/remote/controller.rs]: Agree! Fixed.

File: crates/fluvio-spu/src/mirroring/remote/controller.rs [hunk 4]
```diff
@@ -0,0 +1,523 @@
...
+        self.connect_failure.load(Ordering::Relaxed)
+    }
+}
+
+/// State for mirror controller which can be shared across tasks
+#[derive(Debug)]
+pub(crate) struct MirrorControllerState {
+    metrics: MirrorControllerMetrics,
+
+    #[allow(dead_code)]
```
[the remark this answers is not in the document]
Comment (fraidev) [reply crates/fluvio-spu/src/mirroring/remote/controller.rs]: Fixed!

File: crates/fluvio-spu/src/mirroring/remote/controller.rs [hunk 5]
```diff
@@ -0,0 +1,523 @@
...
+        }
+    }
+
+    #[allow(dead_code)]
+    pub(crate) fn get_metrics(&self) -> &MirrorControllerMetrics {
+        &self.metrics
+    }
+}
+
+const CLUSTER_LOOKUP_SEC: u64 = 5; // 1 sec
```
[the remark this answers is not in the document]
Comment (fraidev) [reply crates/fluvio-spu/src/mirroring/remote/controller.rs]: I forgot this one haha, removed.

File: crates/fluvio-controlplane/src/spu_api/update_mirror.rs (hunk 1 above)
Comment (fraidev) [reply crates/fluvio-controlplane/src/spu_api/update_mirror.rs]: fixed!

**Review by digikata — APPROVED** [review]
LGTM

**Pull request merged**